From 9c01a70642141031d4aadc4c703cb9ea4eb4444b Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 21 Apr 2017 17:02:16 -0700
Subject: [PATCH] SAPI-286 SAPI DeleteInstance should cope with VMAPI DeleteVm
 returning 200 and no job_uuid Reviewed by: Julien Gilli
 <julien.gilli@joyent.com>

---
 lib/server/vmapiplus.js | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/lib/server/vmapiplus.js b/lib/server/vmapiplus.js
index ca0d18e..ee6cfaf 100644
--- a/lib/server/vmapiplus.js
+++ b/lib/server/vmapiplus.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -82,17 +82,29 @@ function deleteVm(uuid, cb) {
     var params = {};
     params.uuid = uuid;
 
-    vmapi.deleteVm(params, function (err, res) {
+    vmapi.deleteVm(params, function (err, body) {
         if (err) {
             log.error(err, 'failed to delete VM %s', uuid);
             return (cb(err));
+        } else if (!body.job_uuid) {
+            /*
+             * For a period of time (between ZAPI-685 and ZAPI-782), VMAPI
+             * DeleteVm could return a "200 OK" with a response body that
+             * did not include a "job_uuid" -- because no destroy job was
+             * created.
+             */
+            log.info({body: body}, 'VMAPI DeleteVm response: no job_uuid, ' +
+                'VM was synchronously marked deleted');
+            cb();
+            return;
         }
 
-        log.info({ job: res.job_uuid }, 'destroy job dispatched');
+        log.info({ vm_uuid: body.vm_uuid, job: body.job_uuid },
+            'destroy job dispatched');
 
         var expected_errors = [ 'ResourceNotFoundError' ];
 
-        waitForJob.call(self, res.job_uuid, expected_errors,
+        waitForJob.call(self, body.job_uuid, expected_errors,
             function (suberr) {
             if (suberr) {
                 suberr = new mod_errors.DestroyFailedError(
-- 
2.21.0

